File: machine-learning-ex1/py/gradientDescent.py
import csv
import numpy as np
import costFunction
import cleandata
import matplotlib.pyplot as plt
from math import fabs

X = [] 
Y = []

with open('ex1data2.txt' , 'r') as csvfile:
    rowreader = csv.reader(csvfile)
    for row in rowreader:
        row.insert(0,1)
        Y.append(row.pop())
        X.append(row)
csvfile.close()

####
alpha = 0.01            #learning rate
m = np.shape(X)[0]   # number of rows
n = np.shape(X)[1]   #number of coloumns
theta = []
EPS = 0.01             #LIMIT {difference in J }
for i in range(0,n):
    theta.append(0)

for i in range(0,m):
    for j in range(0,n):
        X[i][j] = int(X[i][j])
    Y[i] = int(Y[i])
cleandata.clean(X,m,n)
#cleandata.clean(Y,m,1)
#print(X)
#print(Y)
####

J_HISTORY = []
J_HISTORY.append(costFunction.cost(X,Y,theta,m,n))

ctr = 400
while True and ctr > 0:

    theta_temp = []
    costarray = []
    
    for i in range(0,m):
        temp  = []
        for j in range(0,n):
            temp.append(X[i][j])
        costarray.append(costFunction.h(temp , theta , n) - Y[i])

    for i in range(0,n):
        temp = 0
        for j in range(0,m):
            temp += costarray[j]*X[j][i]
        theta_temp.append(theta[i] - (alpha/m)*temp)
    
    theta = theta_temp
    J_HISTORY.append(costFunction.cost(X,Y,theta,m,n))

    if fabs(J_HISTORY[-1] - J_HISTORY[-2]) <= EPS:
        break
    ctr = ctr - 1
print (theta)

plt.plot(J_HISTORY)
plt.show()

File: machine-learning-ex1/R/gradientDescent.R
setwd('/home/vansh/Desktop/machine-learning-by-andrew-ng/machine-learning-ex1/R')
DATA <- read.csv('ex1data2.txt' , header = FALSE , sep = ',')
DATA = as.matrix(DATA)  #idk this had DATA as list and i wasnt able to perform matrix multiplication
# i need to read more into this

### DATA ###
Y <- DATA[,ncol(DATA)]
X <- DATA[,1:ncol(DATA)-1]
alpha = 0.01
n = ncol(X) + 1  #taking account of the 1 coloumn
m = nrow(X)
theta = matrix(0,n,1)
EPS = 0.001
###------###

### applying normalisation ###
for (i in range(1:ncol(X))){
  avg <- mean(X[,i])
  sigma = sd(X[,i])
  X[,i] <- X[,i] - avg
  X[,i] <- X[,i]/sigma
}
###-----------------------###

### adding coloumn of 1 in X ###
X <- cbind(matrix(1,m,1) , X)
###---------------###

costfunction <- function(){
    return ( (1/(2*m)) * ( t(X %*% theta - Y) %*% (X %*% theta - Y) ) )
}
prev <- costfunction()
ctr <- 400
count <- 0
J_HISTORY = c(prev)
repeat{
  theta <- theta - (alpha/m) * ( t(X) %*% (X %*% theta - Y) ) 
  curr <- costfunction()
  if (abs(curr - prev) <= EPS || ctr <= 0){    break
  }
 J_HISTORY <- append(J_HISTORY  , c(curr) , after = length(J_HISTORY))
  ctr <- ctr - 1
  count <- count + 1
  prev <- curr  
}
# theta
# curr
# count 
 plot(J_HISTORY)
# theta[1] + 1650*theta[2] + 3*theta[3]
# 
theta

